In [1]:
import pandas as pd

In [2]:
pd.set_option('display.max_columns', 5000);
pd.set_option('display.width', 1000000);

In [3]:
admissions = pd.read_csv("/Users/liloli/Desktop/6895Project/Carepre/ADMISSIONS.csv")

In [4]:
admissions.head()

ROW_ID  SUBJECT_ID  HADM_ID            ADMITTIME            DISCHTIME DEATHTIME ADMISSION_TYPE         ADMISSION_LOCATION         DISCHARGE_LOCATION INSURANCE LANGUAGE           RELIGION MARITAL_STATUS ETHNICITY            EDREGTIME            EDOUTTIME                                          DIAGNOSIS  HOSPITAL_EXPIRE_FLAG  HAS_CHARTEVENTS_DATA
0      21          22   165315  2196-04-09 12:26:00  2196-04-10 15:54:00       NaN      EMERGENCY       EMERGENCY ROOM ADMIT  DISC-TRAN CANCER/CHLDRN H   Private      NaN       UNOBTAINABLE        MARRIED     WHITE  2196-04-09 10:06:00  2196-04-09 13:24:00                            BENZODIAZEPINE OVERDOSE                     0                     1
1      22          23   152223  2153-09-03 07:15:00  2153-09-08 19:10:00       NaN       ELECTIVE  PHYS REFERRAL/NORMAL DELI           HOME HEALTH CARE  Medicare      NaN           CATHOLIC        MARRIED     WHITE                  NaN                  NaN  CORONARY ARTERY DISEASE\CORONARY ARTERY BYPASS...                     0                     1
2      23          23   124321  2157-10-18 19:34:00  2157-10-25 14:00:00       NaN      EMERGENCY  TRANSFER FROM HOSP/EXTRAM           HOME HEALTH CARE  Medicare     ENGL           CATHOLIC        MARRIED     WHITE                  NaN                  NaN                                         BRAIN MASS                     0                     1
3      24          24   161859  2139-06-06 16:14:00  2139-06-09 12:48:00       NaN      EMERGENCY  TRANSFER FROM HOSP/EXTRAM                       HOME   Private      NaN  PROTESTANT QUAKER         SINGLE     WHITE                  NaN                  NaN                     INTERIOR MYOCARDIAL INFARCTION                     0                     1
4      25          25   129635  2160-11-02 02:06:00  2160-11-05 14:55:00       NaN      EMERGENCY       EMERGENCY ROOM ADMIT                       HOME   Private      NaN       UNOBTAINABLE        MARRIED     WHITE  2160-11-02 01:01:00  2160-11-02 04:27:00                            ACUTE CORONARY SYNDROME                     0                     1

- Display all the data we've got.

In [5]:
ls /Users/liloli/Desktop/6895Project/Carepre

ADMISSIONS.csv        D_ICD_DIAGNOSES.csv   Untitled.ipynb
CAREGIVERS.csv        D_ICD_PROCEDURES.csv  admissions_event.txt
CPTEVENTS.csv         PRESCRIPTIONS.csv
DIAGNOSES_ICD.csv     PROCEDURES_ICD.csv


- Let's clean data for the "diagnose" stage.

In [6]:
diagnose_icd =  pd.read_csv("/Users/liloli/Desktop/6895Project/Carepre/DIAGNOSES_ICD.csv")

In [7]:
diagnose_icd.insert(2, "event_type", "diagnose") 

In [8]:
diagnose_icd.head()

ROW_ID  SUBJECT_ID event_type  HADM_ID  SEQ_NUM ICD9_CODE
0    1297         109   diagnose   172335      1.0     40301
1    1298         109   diagnose   172335      2.0       486
2    1299         109   diagnose   172335      3.0     58281
3    1300         109   diagnose   172335      4.0      5855
4    1301         109   diagnose   172335      5.0      4254

In [9]:
procedures_icd =  pd.read_csv("/Users/liloli/Desktop/6895Project/Carepre/PROCEDURES_ICD.csv")

In [10]:
procedures_icd.insert(2, "event_type", "treatment") 

In [11]:
procedures_icd.head()

ROW_ID  SUBJECT_ID event_type  HADM_ID  SEQ_NUM  ICD9_CODE
0     944       62641  treatment   154460        3       3404
1     945        2592  treatment   130856        1       9671
2     946        2592  treatment   130856        2       3893
3     947       55357  treatment   119355        1       9672
4     948       55357  treatment   119355        2        331

In [12]:
import numpy as np
import pandas as pd

In [13]:
admissions_diagnose = pd.merge(admissions, diagnose_icd, on=["SUBJECT_ID","HADM_ID"], how='left')

In [14]:
admissions_diagnose.head()

ROW_ID_x  SUBJECT_ID  HADM_ID            ADMITTIME            DISCHTIME DEATHTIME ADMISSION_TYPE    ADMISSION_LOCATION         DISCHARGE_LOCATION INSURANCE LANGUAGE      RELIGION MARITAL_STATUS ETHNICITY            EDREGTIME            EDOUTTIME                DIAGNOSIS  HOSPITAL_EXPIRE_FLAG  HAS_CHARTEVENTS_DATA  ROW_ID_y event_type  SEQ_NUM ICD9_CODE
0        21          22   165315  2196-04-09 12:26:00  2196-04-10 15:54:00       NaN      EMERGENCY  EMERGENCY ROOM ADMIT  DISC-TRAN CANCER/CHLDRN H   Private      NaN  UNOBTAINABLE        MARRIED     WHITE  2196-04-09 10:06:00  2196-04-09 13:24:00  BENZODIAZEPINE OVERDOSE                     0                     1       151   diagnose      1.0      9678
1        21          22   165315  2196-04-09 12:26:00  2196-04-10 15:54:00       NaN      EMERGENCY  EMERGENCY ROOM ADMIT  DISC-TRAN CANCER/CHLDRN H   Private      NaN  UNOBTAINABLE        MARRIED     WHITE  2196-04-09 10:06:00  2196-04-09 13:24:00  BENZODIAZEPINE OVERDOSE                     0                     1       152   diagnose      2.0      9693
2        21          22   165315  2196-04-09 12:26:00  2196-04-10 15:54:00       NaN      EMERGENCY  EMERGENCY ROOM ADMIT  DISC-TRAN CANCER/CHLDRN H   Private      NaN  UNOBTAINABLE        MARRIED     WHITE  2196-04-09 10:06:00  2196-04-09 13:24:00  BENZODIAZEPINE OVERDOSE                     0                     1       153   diagnose      3.0     E9502
3        21          22   165315  2196-04-09 12:26:00  2196-04-10 15:54:00       NaN      EMERGENCY  EMERGENCY ROOM ADMIT  DISC-TRAN CANCER/CHLDRN H   Private      NaN  UNOBTAINABLE        MARRIED     WHITE  2196-04-09 10:06:00  2196-04-09 13:24:00  BENZODIAZEPINE OVERDOSE                     0                     1       154   diagnose      4.0     E9503
4        21          22   165315  2196-04-09 12:26:00  2196-04-10 15:54:00       NaN      EMERGENCY  EMERGENCY ROOM ADMIT  DISC-TRAN CANCER/CHLDRN H   Private      NaN  UNOBTAINABLE        MARRIED     WHITE  2196-04-09 10:06:00  2196-04-09 13:24:00  BENZODIAZEPINE OVERDOSE                     0                     1       155   diagnose      5.0      3488

In [15]:
D_ICD_DIAGNOSES = pd.read_csv("/Users/liloli/Desktop/6895Project/Carepre/D_ICD_DIAGNOSES.csv")

In [16]:
D_ICD_DIAGNOSES.head()

ROW_ID ICD9_CODE               SHORT_TITLE                                         LONG_TITLE
0     174     01166     TB pneumonia-oth test  Tuberculous pneumonia [any form], tubercle bac...
1     175     01170    TB pneumothorax-unspec              Tuberculous pneumothorax, unspecified
2     176     01171   TB pneumothorax-no exam  Tuberculous pneumothorax, bacteriological or h...
3     177     01172  TB pneumothorx-exam unkn  Tuberculous pneumothorax, bacteriological or h...
4     178     01173  TB pneumothorax-micro dx  Tuberculous pneumothorax, tubercle bacilli fou...

In [17]:
admission_D_ICD_DIAGNOSES = pd.merge(admissions_diagnose, D_ICD_DIAGNOSES, on=["ICD9_CODE"], how='left')

- `admission_D_ICD_DIAGNOSES` is the table contains all the information we need when "DIAGNOSES", which provides the patient's id, the ICD9_code of the disease and the description(`LONG_TITLE`) of this disease(according to the specific ICD9_code).

In [18]:
admission_D_ICD_DIAGNOSES.head()

ROW_ID_x  SUBJECT_ID  HADM_ID            ADMITTIME            DISCHTIME DEATHTIME ADMISSION_TYPE    ADMISSION_LOCATION         DISCHARGE_LOCATION INSURANCE LANGUAGE      RELIGION MARITAL_STATUS ETHNICITY            EDREGTIME            EDOUTTIME                DIAGNOSIS  HOSPITAL_EXPIRE_FLAG  HAS_CHARTEVENTS_DATA  ROW_ID_y event_type  SEQ_NUM ICD9_CODE   ROW_ID               SHORT_TITLE                                         LONG_TITLE
0        21          22   165315  2196-04-09 12:26:00  2196-04-10 15:54:00       NaN      EMERGENCY  EMERGENCY ROOM ADMIT  DISC-TRAN CANCER/CHLDRN H   Private      NaN  UNOBTAINABLE        MARRIED     WHITE  2196-04-09 10:06:00  2196-04-09 13:24:00  BENZODIAZEPINE OVERDOSE                     0                     1       151   diagnose      1.0      9678  11344.0  Pois-sedative/hypnot NEC         Poisoning by other sedatives and hypnotics
1        21          22   165315  2196-04-09 12:26:00  2196-04-10 15:54:00       NaN      EMERGENCY  EMERGENCY ROOM ADMIT  DISC-TRAN CANCER/CHLDRN H   Private      NaN  UNOBTAINABLE        MARRIED     WHITE  2196-04-09 10:06:00  2196-04-09 13:24:00  BENZODIAZEPINE OVERDOSE                     0                     1       152   diagnose      2.0      9693  11553.0  Poison-antipsychotic NEC  Poisoning by other antipsychotics, neuroleptic...
2        21          22   165315  2196-04-09 12:26:00  2196-04-10 15:54:00       NaN      EMERGENCY  EMERGENCY ROOM ADMIT  DISC-TRAN CANCER/CHLDRN H   Private      NaN  UNOBTAINABLE        MARRIED     WHITE  2196-04-09 10:06:00  2196-04-09 13:24:00  BENZODIAZEPINE OVERDOSE                     0                     1       153   diagnose      3.0     E9502   9947.0     Poison-sedat/hypnotic  Suicide and self-inflicted poisoning by other ...
3        21          22   165315  2196-04-09 12:26:00  2196-04-10 15:54:00       NaN      EMERGENCY  EMERGENCY ROOM ADMIT  DISC-TRAN CANCER/CHLDRN H   Private      NaN  UNOBTAINABLE        MARRIED     WHITE  2196-04-09 10:06:00  2196-04-09 13:24:00  BENZODIAZEPINE OVERDOSE                     0                     1       154   diagnose      4.0     E9503   9948.0   Poison-psychotropic agt  Suicide and self-inflicted poisoning by tranqu...
4        21          22   165315  2196-04-09 12:26:00  2196-04-10 15:54:00       NaN      EMERGENCY  EMERGENCY ROOM ADMIT  DISC-TRAN CANCER/CHLDRN H   Private      NaN  UNOBTAINABLE        MARRIED     WHITE  2196-04-09 10:06:00  2196-04-09 13:24:00  BENZODIAZEPINE OVERDOSE                     0                     1       155   diagnose      5.0      3488      NaN                       NaN                                                NaN

- Let's clean data for the "treatment" stage.

In [19]:
admissions_procedures  = pd.merge(admissions, procedures_icd, on=["SUBJECT_ID","HADM_ID"], how='left')

In [20]:
admissions_procedures.head()

ROW_ID_x  SUBJECT_ID  HADM_ID            ADMITTIME            DISCHTIME DEATHTIME ADMISSION_TYPE         ADMISSION_LOCATION         DISCHARGE_LOCATION INSURANCE LANGUAGE      RELIGION MARITAL_STATUS ETHNICITY            EDREGTIME            EDOUTTIME                                          DIAGNOSIS  HOSPITAL_EXPIRE_FLAG  HAS_CHARTEVENTS_DATA  ROW_ID_y event_type  SEQ_NUM  ICD9_CODE
0        21          22   165315  2196-04-09 12:26:00  2196-04-10 15:54:00       NaN      EMERGENCY       EMERGENCY ROOM ADMIT  DISC-TRAN CANCER/CHLDRN H   Private      NaN  UNOBTAINABLE        MARRIED     WHITE  2196-04-09 10:06:00  2196-04-09 13:24:00                            BENZODIAZEPINE OVERDOSE                     0                     1    9894.0  treatment      1.0     9604.0
1        21          22   165315  2196-04-09 12:26:00  2196-04-10 15:54:00       NaN      EMERGENCY       EMERGENCY ROOM ADMIT  DISC-TRAN CANCER/CHLDRN H   Private      NaN  UNOBTAINABLE        MARRIED     WHITE  2196-04-09 10:06:00  2196-04-09 13:24:00                            BENZODIAZEPINE OVERDOSE                     0                     1    9895.0  treatment      2.0     9671.0
2        21          22   165315  2196-04-09 12:26:00  2196-04-10 15:54:00       NaN      EMERGENCY       EMERGENCY ROOM ADMIT  DISC-TRAN CANCER/CHLDRN H   Private      NaN  UNOBTAINABLE        MARRIED     WHITE  2196-04-09 10:06:00  2196-04-09 13:24:00                            BENZODIAZEPINE OVERDOSE                     0                     1    9896.0  treatment      3.0     9634.0
3        22          23   152223  2153-09-03 07:15:00  2153-09-08 19:10:00       NaN       ELECTIVE  PHYS REFERRAL/NORMAL DELI           HOME HEALTH CARE  Medicare      NaN      CATHOLIC        MARRIED     WHITE                  NaN                  NaN  CORONARY ARTERY DISEASE\CORONARY ARTERY BYPASS...                     0                     1  105252.0  treatment      1.0     3613.0
4        22          23   152223  2153-09-03 07:15:00  2153-09-08 19:10:00       NaN       ELECTIVE  PHYS REFERRAL/NORMAL DELI           HOME HEALTH CARE  Medicare      NaN      CATHOLIC        MARRIED     WHITE                  NaN                  NaN  CORONARY ARTERY DISEASE\CORONARY ARTERY BYPASS...                     0                     1  105253.0  treatment      2.0     3615.0

In [21]:
D_ICD_PROCEDURES = pd.read_csv("/Users/liloli/Desktop/6895Project/Carepre/D_ICD_PROCEDURES.csv")

In [22]:
D_ICD_PROCEDURES.head()

ROW_ID  ICD9_CODE               SHORT_TITLE                                         LONG_TITLE
0     264        851                Canthotomy                                         Canthotomy
1     265        852           Blepharorrhaphy                                    Blepharorrhaphy
2     266        859   Adjust lid position NEC                   Other adjustment of lid position
3     267        861  Lid reconst w skin graft   Reconstruction of eyelid with skin flap or graft
4     268        862   Lid reconst w muc graft  Reconstruction of eyelid with mucous membrane ...

In [23]:
admissions_D_ICD_PROCEDURES = pd.merge(admissions_procedures, D_ICD_PROCEDURES, on=["ICD9_CODE"], how='left')

In [24]:
admissions_D_ICD_PROCEDURES.head()

ROW_ID_x  SUBJECT_ID  HADM_ID            ADMITTIME            DISCHTIME DEATHTIME ADMISSION_TYPE         ADMISSION_LOCATION         DISCHARGE_LOCATION INSURANCE LANGUAGE      RELIGION MARITAL_STATUS ETHNICITY            EDREGTIME            EDOUTTIME                                          DIAGNOSIS  HOSPITAL_EXPIRE_FLAG  HAS_CHARTEVENTS_DATA  ROW_ID_y event_type  SEQ_NUM  ICD9_CODE  ROW_ID               SHORT_TITLE                                         LONG_TITLE
0        21          22   165315  2196-04-09 12:26:00  2196-04-10 15:54:00       NaN      EMERGENCY       EMERGENCY ROOM ADMIT  DISC-TRAN CANCER/CHLDRN H   Private      NaN  UNOBTAINABLE        MARRIED     WHITE  2196-04-09 10:06:00  2196-04-09 13:24:00                            BENZODIAZEPINE OVERDOSE                     0                     1    9894.0  treatment      1.0     9604.0  3607.0  Insert endotracheal tube                     Insertion of endotracheal tube
1        21          22   165315  2196-04-09 12:26:00  2196-04-10 15:54:00       NaN      EMERGENCY       EMERGENCY ROOM ADMIT  DISC-TRAN CANCER/CHLDRN H   Private      NaN  UNOBTAINABLE        MARRIED     WHITE  2196-04-09 10:06:00  2196-04-09 13:24:00                            BENZODIAZEPINE OVERDOSE                     0                     1    9895.0  treatment      2.0     9671.0  3678.0  Cont inv mec ven <96 hrs  Continuous invasive mechanical ventilation for...
2        21          22   165315  2196-04-09 12:26:00  2196-04-10 15:54:00       NaN      EMERGENCY       EMERGENCY ROOM ADMIT  DISC-TRAN CANCER/CHLDRN H   Private      NaN  UNOBTAINABLE        MARRIED     WHITE  2196-04-09 10:06:00  2196-04-09 13:24:00                            BENZODIAZEPINE OVERDOSE                     0                     1    9896.0  treatment      3.0     9634.0  3632.0  Gastric tube irrigat NEC            Other irrigation of (naso-)gastric tube
3        22          23   152223  2153-09-03 07:15:00  2153-09-08 19:10:00       NaN       ELECTIVE  PHYS REFERRAL/NORMAL DELI           HOME HEALTH CARE  Medicare      NaN      CATHOLIC        MARRIED     WHITE                  NaN                  NaN  CORONARY ARTERY DISEASE\CORONARY ARTERY BYPASS...                     0                     1  105252.0  treatment      1.0     3613.0   663.0  Aortocor bypas-3 cor art  (Aorto)coronary bypass of three coronary arteries
4        22          23   152223  2153-09-03 07:15:00  2153-09-08 19:10:00       NaN       ELECTIVE  PHYS REFERRAL/NORMAL DELI           HOME HEALTH CARE  Medicare      NaN      CATHOLIC        MARRIED     WHITE                  NaN                  NaN  CORONARY ARTERY DISEASE\CORONARY ARTERY BYPASS...                     0                     1  105253.0  treatment      2.0     3615.0   665.0  1 int mam-cor art bypass     Single internal mammary-coronary artery bypass

Each row of this table contains a unique `HADM_ID`, which represents a single patient’s admission to the hospital. `HADM_ID` ranges from 1000000 - 1999999. It is possible for this table to have duplicate `SUBJECT_ID`, indicating that a single patient had multiple admissions to the hospital. The ADMISSIONS table can be linked to the PATIENTS table using `SUBJECT_ID`

2 important index for a patient:
- `SUBJECT_ID`: patient's id
- `HADM_ID`: a single patient’s admission to the hospital

Both DIAGNOSES and TREATMENT procedures are defined as medical events. So we concat 2 tables. `admissions_event` dataframe is the major data frame we need for future analysis.

In [25]:
admissions_event = pd.concat([admissions_D_ICD_PROCEDURES, admission_D_ICD_DIAGNOSES], axis=0)

In [26]:
admissions_event.head()

ROW_ID_x  SUBJECT_ID  HADM_ID            ADMITTIME            DISCHTIME DEATHTIME ADMISSION_TYPE         ADMISSION_LOCATION         DISCHARGE_LOCATION INSURANCE LANGUAGE      RELIGION MARITAL_STATUS ETHNICITY            EDREGTIME            EDOUTTIME                                          DIAGNOSIS  HOSPITAL_EXPIRE_FLAG  HAS_CHARTEVENTS_DATA  ROW_ID_y event_type  SEQ_NUM ICD9_CODE  ROW_ID               SHORT_TITLE                                         LONG_TITLE
0        21          22   165315  2196-04-09 12:26:00  2196-04-10 15:54:00       NaN      EMERGENCY       EMERGENCY ROOM ADMIT  DISC-TRAN CANCER/CHLDRN H   Private      NaN  UNOBTAINABLE        MARRIED     WHITE  2196-04-09 10:06:00  2196-04-09 13:24:00                            BENZODIAZEPINE OVERDOSE                     0                     1    9894.0  treatment      1.0      9604  3607.0  Insert endotracheal tube                     Insertion of endotracheal tube
1        21          22   165315  2196-04-09 12:26:00  2196-04-10 15:54:00       NaN      EMERGENCY       EMERGENCY ROOM ADMIT  DISC-TRAN CANCER/CHLDRN H   Private      NaN  UNOBTAINABLE        MARRIED     WHITE  2196-04-09 10:06:00  2196-04-09 13:24:00                            BENZODIAZEPINE OVERDOSE                     0                     1    9895.0  treatment      2.0      9671  3678.0  Cont inv mec ven <96 hrs  Continuous invasive mechanical ventilation for...
2        21          22   165315  2196-04-09 12:26:00  2196-04-10 15:54:00       NaN      EMERGENCY       EMERGENCY ROOM ADMIT  DISC-TRAN CANCER/CHLDRN H   Private      NaN  UNOBTAINABLE        MARRIED     WHITE  2196-04-09 10:06:00  2196-04-09 13:24:00                            BENZODIAZEPINE OVERDOSE                     0                     1    9896.0  treatment      3.0      9634  3632.0  Gastric tube irrigat NEC            Other irrigation of (naso-)gastric tube
3        22          23   152223  2153-09-03 07:15:00  2153-09-08 19:10:00       NaN       ELECTIVE  PHYS REFERRAL/NORMAL DELI           HOME HEALTH CARE  Medicare      NaN      CATHOLIC        MARRIED     WHITE                  NaN                  NaN  CORONARY ARTERY DISEASE\CORONARY ARTERY BYPASS...                     0                     1  105252.0  treatment      1.0      3613   663.0  Aortocor bypas-3 cor art  (Aorto)coronary bypass of three coronary arteries
4        22          23   152223  2153-09-03 07:15:00  2153-09-08 19:10:00       NaN       ELECTIVE  PHYS REFERRAL/NORMAL DELI           HOME HEALTH CARE  Medicare      NaN      CATHOLIC        MARRIED     WHITE                  NaN                  NaN  CORONARY ARTERY DISEASE\CORONARY ARTERY BYPASS...                     0                     1  105253.0  treatment      2.0      3615   665.0  1 int mam-cor art bypass     Single internal mammary-coronary artery bypass

Next step: embedding using __NLP__ algorithms. 

Basic idea: each row represents a clinical/medical treatment/diagnose for a patient. We want to know the relationship between patients with the similar diseases which need similar treatments and have similar syptoms. Using __machine learning__ algorithms seems a good way to identify and cluster these similarities. However, the treatment/diagnose descriptions are written in English, which cannot be understood by computer. So we need __NLP__ algorithms to help us to convert English to numeric vectors.

Genism word2vec requires that a format of ‘list of lists’ for training where every document is contained in a list and every list contains lists of tokens of that document. At first, we need to generate a format of ‘list of lists’ for training the make model word embedding. To be more specific, each make model is contained in a list and every list contains lists of features of that make model.

To achieve this, we need to do the following things :

## Step A. Create a new column for Make Model

In [27]:
df = admissions_event   #for coding convienence. Use df to refer to admissions_event.

In [28]:
df.head()

ROW_ID_x  SUBJECT_ID  HADM_ID            ADMITTIME            DISCHTIME DEATHTIME ADMISSION_TYPE         ADMISSION_LOCATION         DISCHARGE_LOCATION INSURANCE LANGUAGE      RELIGION MARITAL_STATUS ETHNICITY            EDREGTIME            EDOUTTIME                                          DIAGNOSIS  HOSPITAL_EXPIRE_FLAG  HAS_CHARTEVENTS_DATA  ROW_ID_y event_type  SEQ_NUM ICD9_CODE  ROW_ID               SHORT_TITLE                                         LONG_TITLE
0        21          22   165315  2196-04-09 12:26:00  2196-04-10 15:54:00       NaN      EMERGENCY       EMERGENCY ROOM ADMIT  DISC-TRAN CANCER/CHLDRN H   Private      NaN  UNOBTAINABLE        MARRIED     WHITE  2196-04-09 10:06:00  2196-04-09 13:24:00                            BENZODIAZEPINE OVERDOSE                     0                     1    9894.0  treatment      1.0      9604  3607.0  Insert endotracheal tube                     Insertion of endotracheal tube
1        21          22   165315  2196-04-09 12:26:00  2196-04-10 15:54:00       NaN      EMERGENCY       EMERGENCY ROOM ADMIT  DISC-TRAN CANCER/CHLDRN H   Private      NaN  UNOBTAINABLE        MARRIED     WHITE  2196-04-09 10:06:00  2196-04-09 13:24:00                            BENZODIAZEPINE OVERDOSE                     0                     1    9895.0  treatment      2.0      9671  3678.0  Cont inv mec ven <96 hrs  Continuous invasive mechanical ventilation for...
2        21          22   165315  2196-04-09 12:26:00  2196-04-10 15:54:00       NaN      EMERGENCY       EMERGENCY ROOM ADMIT  DISC-TRAN CANCER/CHLDRN H   Private      NaN  UNOBTAINABLE        MARRIED     WHITE  2196-04-09 10:06:00  2196-04-09 13:24:00                            BENZODIAZEPINE OVERDOSE                     0                     1    9896.0  treatment      3.0      9634  3632.0  Gastric tube irrigat NEC            Other irrigation of (naso-)gastric tube
3        22          23   152223  2153-09-03 07:15:00  2153-09-08 19:10:00       NaN       ELECTIVE  PHYS REFERRAL/NORMAL DELI           HOME HEALTH CARE  Medicare      NaN      CATHOLIC        MARRIED     WHITE                  NaN                  NaN  CORONARY ARTERY DISEASE\CORONARY ARTERY BYPASS...                     0                     1  105252.0  treatment      1.0      3613   663.0  Aortocor bypas-3 cor art  (Aorto)coronary bypass of three coronary arteries
4        22          23   152223  2153-09-03 07:15:00  2153-09-08 19:10:00       NaN       ELECTIVE  PHYS REFERRAL/NORMAL DELI           HOME HEALTH CARE  Medicare      NaN      CATHOLIC        MARRIED     WHITE                  NaN                  NaN  CORONARY ARTERY DISEASE\CORONARY ARTERY BYPASS...                     0                     1  105253.0  treatment      2.0      3615   665.0  1 int mam-cor art bypass     Single internal mammary-coronary artery bypass

In [29]:
df['SEQ_NUM_str'] = df['SEQ_NUM'].astype(str)

In [30]:
df['CLINICAL_RECORD']= df['event_type'] + " "+df['SHORT_TITLE']+ " " + df['SEQ_NUM_str']

In [31]:
df

ROW_ID_x  SUBJECT_ID  HADM_ID            ADMITTIME            DISCHTIME DEATHTIME ADMISSION_TYPE         ADMISSION_LOCATION         DISCHARGE_LOCATION INSURANCE LANGUAGE      RELIGION MARITAL_STATUS ETHNICITY            EDREGTIME            EDOUTTIME                                          DIAGNOSIS  HOSPITAL_EXPIRE_FLAG  HAS_CHARTEVENTS_DATA  ROW_ID_y event_type  SEQ_NUM ICD9_CODE   ROW_ID               SHORT_TITLE                                         LONG_TITLE SEQ_NUM_str                         CLINICAL_RECORD
0             21          22   165315  2196-04-09 12:26:00  2196-04-10 15:54:00       NaN      EMERGENCY       EMERGENCY ROOM ADMIT  DISC-TRAN CANCER/CHLDRN H   Private      NaN  UNOBTAINABLE        MARRIED     WHITE  2196-04-09 10:06:00  2196-04-09 13:24:00                            BENZODIAZEPINE OVERDOSE                     0                     1    9894.0  treatment      1.0      9604   3607.0  Insert endotracheal tube                     Insertion of endotracheal tube         1.0  treatment Insert endotracheal tube 1.0
1             21          22   165315  2196-04-09 12:26:00  2196-04-10 15:54:00       NaN      EMERGENCY       EMERGENCY ROOM ADMIT  DISC-TRAN CANCER/CHLDRN H   Private      NaN  UNOBTAINABLE        MARRIED     WHITE  2196-04-09 10:06:00  2196-04-09 13:24:00                            BENZODIAZEPINE OVERDOSE                     0                     1    9895.0  treatment      2.0      9671   3678.0  Cont inv mec ven <96 hrs  Continuous invasive mechanical ventilation for...         2.0  treatment Cont inv mec ven <96 hrs 2.0
2             21          22   165315  2196-04-09 12:26:00  2196-04-10 15:54:00       NaN      EMERGENCY       EMERGENCY ROOM ADMIT  DISC-TRAN CANCER/CHLDRN H   Private      NaN  UNOBTAINABLE        MARRIED     WHITE  2196-04-09 10:06:00  2196-04-09 13:24:00                            BENZODIAZEPINE OVERDOSE                     0                     1    9896.0  treatment      3.0      9634   3632.0  Gastric tube irrigat NEC            Other irrigation of (naso-)gastric tube         3.0  treatment Gastric tube irrigat NEC 3.0
3             22          23   152223  2153-09-03 07:15:00  2153-09-08 19:10:00       NaN       ELECTIVE  PHYS REFERRAL/NORMAL DELI           HOME HEALTH CARE  Medicare      NaN      CATHOLIC        MARRIED     WHITE                  NaN                  NaN  CORONARY ARTERY DISEASE\CORONARY ARTERY BYPASS...                     0                     1  105252.0  treatment      1.0      3613    663.0  Aortocor bypas-3 cor art  (Aorto)coronary bypass of three coronary arteries         1.0  treatment Aortocor bypas-3 cor art 1.0
4             22          23   152223  2153-09-03 07:15:00  2153-09-08 19:10:00       NaN       ELECTIVE  PHYS REFERRAL/NORMAL DELI           HOME HEALTH CARE  Medicare      NaN      CATHOLIC        MARRIED     WHITE                  NaN                  NaN  CORONARY ARTERY DISEASE\CORONARY ARTERY BYPASS...                     0                     1  105253.0  treatment      2.0      3615    665.0  1 int mam-cor art bypass     Single internal mammary-coronary artery bypass         2.0  treatment 1 int mam-cor art bypass 2.0
...          ...         ...      ...                  ...                  ...       ...            ...                        ...                        ...       ...      ...           ...            ...       ...                  ...                  ...                                                ...                   ...                   ...       ...        ...      ...       ...      ...                       ...                                                ...         ...                                     ...
651042     58598       98813   190264  2131-10-25 03:09:00  2131-10-26 17:44:00       NaN      EMERGENCY  CLINIC REFERRAL/PREMATURE                       HOME   Private     ENGL      CATHOLIC        MARRIED     WHITE  2131-10-25 00:08:00  2131-10-25 04:35:00                          

`CLINICAL_RECORD` is the overall literal description that contains all the information for one row (one record). Let's apply __word2vec__ algorithm to this column. 

In [32]:
df['CLINICAL_RECORD'].fillna("None", inplace = True)
df

ROW_ID_x  SUBJECT_ID  HADM_ID            ADMITTIME            DISCHTIME DEATHTIME ADMISSION_TYPE         ADMISSION_LOCATION         DISCHARGE_LOCATION INSURANCE LANGUAGE      RELIGION MARITAL_STATUS ETHNICITY            EDREGTIME            EDOUTTIME                                          DIAGNOSIS  HOSPITAL_EXPIRE_FLAG  HAS_CHARTEVENTS_DATA  ROW_ID_y event_type  SEQ_NUM ICD9_CODE   ROW_ID               SHORT_TITLE                                         LONG_TITLE SEQ_NUM_str                         CLINICAL_RECORD
0             21          22   165315  2196-04-09 12:26:00  2196-04-10 15:54:00       NaN      EMERGENCY       EMERGENCY ROOM ADMIT  DISC-TRAN CANCER/CHLDRN H   Private      NaN  UNOBTAINABLE        MARRIED     WHITE  2196-04-09 10:06:00  2196-04-09 13:24:00                            BENZODIAZEPINE OVERDOSE                     0                     1    9894.0  treatment      1.0      9604   3607.0  Insert endotracheal tube                     Insertion of endotracheal tube         1.0  treatment Insert endotracheal tube 1.0
1             21          22   165315  2196-04-09 12:26:00  2196-04-10 15:54:00       NaN      EMERGENCY       EMERGENCY ROOM ADMIT  DISC-TRAN CANCER/CHLDRN H   Private      NaN  UNOBTAINABLE        MARRIED     WHITE  2196-04-09 10:06:00  2196-04-09 13:24:00                            BENZODIAZEPINE OVERDOSE                     0                     1    9895.0  treatment      2.0      9671   3678.0  Cont inv mec ven <96 hrs  Continuous invasive mechanical ventilation for...         2.0  treatment Cont inv mec ven <96 hrs 2.0
2             21          22   165315  2196-04-09 12:26:00  2196-04-10 15:54:00       NaN      EMERGENCY       EMERGENCY ROOM ADMIT  DISC-TRAN CANCER/CHLDRN H   Private      NaN  UNOBTAINABLE        MARRIED     WHITE  2196-04-09 10:06:00  2196-04-09 13:24:00                            BENZODIAZEPINE OVERDOSE                     0                     1    9896.0  treatment      3.0      9634   3632.0  Gastric tube irrigat NEC            Other irrigation of (naso-)gastric tube         3.0  treatment Gastric tube irrigat NEC 3.0
3             22          23   152223  2153-09-03 07:15:00  2153-09-08 19:10:00       NaN       ELECTIVE  PHYS REFERRAL/NORMAL DELI           HOME HEALTH CARE  Medicare      NaN      CATHOLIC        MARRIED     WHITE                  NaN                  NaN  CORONARY ARTERY DISEASE\CORONARY ARTERY BYPASS...                     0                     1  105252.0  treatment      1.0      3613    663.0  Aortocor bypas-3 cor art  (Aorto)coronary bypass of three coronary arteries         1.0  treatment Aortocor bypas-3 cor art 1.0
4             22          23   152223  2153-09-03 07:15:00  2153-09-08 19:10:00       NaN       ELECTIVE  PHYS REFERRAL/NORMAL DELI           HOME HEALTH CARE  Medicare      NaN      CATHOLIC        MARRIED     WHITE                  NaN                  NaN  CORONARY ARTERY DISEASE\CORONARY ARTERY BYPASS...                     0                     1  105253.0  treatment      2.0      3615    665.0  1 int mam-cor art bypass     Single internal mammary-coronary artery bypass         2.0  treatment 1 int mam-cor art bypass 2.0
...          ...         ...      ...                  ...                  ...       ...            ...                        ...                        ...       ...      ...           ...            ...       ...                  ...                  ...                                                ...                   ...                   ...       ...        ...      ...       ...      ...                       ...                                                ...         ...                                     ...
651042     58598       98813   190264  2131-10-25 03:09:00  2131-10-26 17:44:00       NaN      EMERGENCY  CLINIC REFERRAL/PREMATURE                       HOME   Private     ENGL      CATHOLIC        MARRIED     WHITE  2131-10-25 00:08:00  2131-10-25 04:35:00                          

## Step b. Generate a format of ‘ list of lists’ for each Make Model with the following features: Engine Fuel Type, Transmission Type, Driven_Wheels, Market Category, Vehicle Size, Vehicle Style

In [33]:
# Select features from original dataset to form a new dataframe 
df1 = df[['ADMISSION_TYPE', 'ETHNICITY', 'DIAGNOSIS', 'LONG_TITLE', 'MARITAL_STATUS', 'INSURANCE', 'CLINICAL_RECORD']]

In [34]:
# For each row, combine all the columns into one column
df2 = df1.apply(lambda x: '$'.join(x.astype(str)), axis=1)

In [35]:
# Store them in a pandas dataframe
df_clean = pd.DataFrame({'clean': df2})

In [36]:
# Create the list of list format of the custom corpus for gensim modeling 
sent = [row.split('$') for row in df_clean['clean']]

In [38]:
# show the example of list of list format of the custom corpus for gensim modeling
sent[:2]

[['EMERGENCY',
  'WHITE',
  'BENZODIAZEPINE OVERDOSE',
  'Insertion of endotracheal tube',
  'MARRIED',
  'Private',
  'treatment Insert endotracheal tube 1.0'],
 ['EMERGENCY',
  'WHITE',
  'BENZODIAZEPINE OVERDOSE',
  'Continuous invasive mechanical ventilation for less than 96 consecutive hours',
  'MARRIED',
  'Private',
  'treatment Cont inv mec ven <96 hrs 2.0']]

## Genism word2vec Model Training

We can train the genism word2vec model with our own custom corpus as following:

In [39]:
import warnings
warnings.filterwarnings('ignore')

In [40]:
import gensim

In [41]:
from gensim.models import Word2Vec

Explain model:
- size: The number of dimensions of the embeddings and the default is 100.
- window: The maximum distance between a target word and words around the target word. The default window is 5. Since our text is shorter, we set window as 3.
- min_count: The minimum count of words to consider when training the model; words with occurrence less than this count will be ignored. The default for min_count is 5.
- workers: The number of partitions during training and the default workers is 3.
- sg: The training algorithm, either CBOW(0) or skip gram(1). The default training algorithm is CBOW. We use skip gram here. 

Train the model:

In [42]:
model = Word2Vec(sent, min_count=1,size= 50,workers=3, window =3, sg = 1)

After training the word2vec model, we can obtain the word embedding directly from the training model as following.

### Elementary tests 

In [43]:
example1 = model['treatment Hemodialysis 1.0']
example1 

array([-0.23486438,  1.3099189 ,  0.26933685, -0.4017835 , -0.18539986,
        1.0577146 ,  1.7902732 ,  1.0384192 ,  0.20688422, -0.06425183,
       -0.3750995 ,  0.03832003,  0.35323018,  0.37527797,  0.40227988,
        0.02382597, -0.35746697,  1.162224  ,  0.15521216, -1.135829  ,
        0.09080667, -0.45336977,  0.8667915 ,  0.7220963 ,  0.0914274 ,
       -0.62333316,  0.07237124, -0.89696705, -1.488926  ,  0.439589  ,
       -0.13379338, -1.4197458 ,  0.19169599,  0.09359043,  0.7757766 ,
        1.7726172 , -0.7100077 ,  0.09100271,  0.86581343,  0.29334748,
        0.65048933,  0.14739047, -0.40889314,  0.35560393,  0.19859987,
        0.8729145 ,  0.7629562 , -0.21866192,  0.54876566,  0.19957252],
      dtype=float32)

In [44]:
example2 = model['treatment Cardiopulm resuscita NOS 2.0']
example2

array([-0.6816129 ,  0.47910285,  0.35045418, -0.6416862 ,  0.07915942,
        0.6016631 ,  0.6219659 ,  0.25610647,  0.29156318,  0.18289956,
       -0.02709882,  0.13591804,  0.63873994, -0.06087818,  0.15216285,
       -0.0848062 , -0.10151227,  0.25353178,  0.15046516, -0.24472442,
        0.16915973, -0.04240311,  0.17592573,  1.0168449 , -0.43944323,
       -0.07466871, -0.09497924, -0.49808767, -0.65476155,  0.4531533 ,
       -0.33136985, -0.36536488, -0.16187024,  0.581411  ,  0.08081025,
        0.97995615, -0.37769514, -0.03873713,  0.5783633 ,  0.13130863,
        0.0212591 , -0.48299986, -0.24238864, -0.33447227,  0.11210129,
        0.46336102,  0.12418678,  0.28111494, -0.09820311,  0.2785122 ],
      dtype=float32)

Perform arithmetic?

In [45]:
example1 * example2

array([ 0.1600866 ,  0.6275859 ,  0.09439023,  0.25781894, -0.01467615,
        0.6363878 ,  1.1134888 ,  0.26594588,  0.06031982, -0.01175163,
        0.01016475,  0.00520838,  0.22562222, -0.02284624,  0.06121206,
       -0.00202059,  0.03628728,  0.29466075,  0.02335402,  0.2779651 ,
        0.01536083,  0.01922429,  0.15249093,  0.73425996, -0.04017715,
        0.04654348, -0.00687377,  0.44676822,  0.97489154,  0.19920121,
        0.04433509,  0.5187253 , -0.03102988,  0.05441451,  0.06269071,
        1.7370871 ,  0.26816648, -0.00352518,  0.5007547 ,  0.03851905,
        0.01382882, -0.07118957,  0.09911105, -0.11893965,  0.0222633 ,
        0.40447456,  0.09474908, -0.06146913, -0.05389049,  0.05558338],
      dtype=float32)

Test 2 treatment for one patient.

In [146]:
model.similarity('treatment Cardiopulm resuscita NOS 2.0','treatment Venous cath NEC 5.0')

0.5385792

In [47]:
df

ROW_ID_x  SUBJECT_ID  HADM_ID            ADMITTIME            DISCHTIME DEATHTIME ADMISSION_TYPE         ADMISSION_LOCATION         DISCHARGE_LOCATION INSURANCE LANGUAGE      RELIGION MARITAL_STATUS ETHNICITY            EDREGTIME            EDOUTTIME                                          DIAGNOSIS  HOSPITAL_EXPIRE_FLAG  HAS_CHARTEVENTS_DATA  ROW_ID_y event_type  SEQ_NUM ICD9_CODE   ROW_ID               SHORT_TITLE                                         LONG_TITLE SEQ_NUM_str                         CLINICAL_RECORD
0             21          22   165315  2196-04-09 12:26:00  2196-04-10 15:54:00       NaN      EMERGENCY       EMERGENCY ROOM ADMIT  DISC-TRAN CANCER/CHLDRN H   Private      NaN  UNOBTAINABLE        MARRIED     WHITE  2196-04-09 10:06:00  2196-04-09 13:24:00                            BENZODIAZEPINE OVERDOSE                     0                     1    9894.0  treatment      1.0      9604   3607.0  Insert endotracheal tube                     Insertion of endotracheal tube         1.0  treatment Insert endotracheal tube 1.0
1             21          22   165315  2196-04-09 12:26:00  2196-04-10 15:54:00       NaN      EMERGENCY       EMERGENCY ROOM ADMIT  DISC-TRAN CANCER/CHLDRN H   Private      NaN  UNOBTAINABLE        MARRIED     WHITE  2196-04-09 10:06:00  2196-04-09 13:24:00                            BENZODIAZEPINE OVERDOSE                     0                     1    9895.0  treatment      2.0      9671   3678.0  Cont inv mec ven <96 hrs  Continuous invasive mechanical ventilation for...         2.0  treatment Cont inv mec ven <96 hrs 2.0
2             21          22   165315  2196-04-09 12:26:00  2196-04-10 15:54:00       NaN      EMERGENCY       EMERGENCY ROOM ADMIT  DISC-TRAN CANCER/CHLDRN H   Private      NaN  UNOBTAINABLE        MARRIED     WHITE  2196-04-09 10:06:00  2196-04-09 13:24:00                            BENZODIAZEPINE OVERDOSE                     0                     1    9896.0  treatment      3.0      9634   3632.0  Gastric tube irrigat NEC            Other irrigation of (naso-)gastric tube         3.0  treatment Gastric tube irrigat NEC 3.0
3             22          23   152223  2153-09-03 07:15:00  2153-09-08 19:10:00       NaN       ELECTIVE  PHYS REFERRAL/NORMAL DELI           HOME HEALTH CARE  Medicare      NaN      CATHOLIC        MARRIED     WHITE                  NaN                  NaN  CORONARY ARTERY DISEASE\CORONARY ARTERY BYPASS...                     0                     1  105252.0  treatment      1.0      3613    663.0  Aortocor bypas-3 cor art  (Aorto)coronary bypass of three coronary arteries         1.0  treatment Aortocor bypas-3 cor art 1.0
4             22          23   152223  2153-09-03 07:15:00  2153-09-08 19:10:00       NaN       ELECTIVE  PHYS REFERRAL/NORMAL DELI           HOME HEALTH CARE  Medicare      NaN      CATHOLIC        MARRIED     WHITE                  NaN                  NaN  CORONARY ARTERY DISEASE\CORONARY ARTERY BYPASS...                     0                     1  105253.0  treatment      2.0      3615    665.0  1 int mam-cor art bypass     Single internal mammary-coronary artery bypass         2.0  treatment 1 int mam-cor art bypass 2.0
...          ...         ...      ...                  ...                  ...       ...            ...                        ...                        ...       ...      ...           ...            ...       ...                  ...                  ...                                                ...                   ...                   ...       ...        ...      ...       ...      ...                       ...                                                ...         ...                                     ...
651042     58598       98813   190264  2131-10-25 03:09:00  2131-10-26 17:44:00       NaN      EMERGENCY  CLINIC REFERRAL/PREMATURE                       HOME   Private     ENGL      CATHOLIC        MARRIED     WHITE  2131-10-25 00:08:00  2131-10-25 04:35:00                          

Now, every clinical record has been mapped to a numeric vector `RECORD_VECTOR`

In [48]:
df['RECORD_VECTOR'] = list(map(lambda x: model[x], df['CLINICAL_RECORD']))

In [49]:
df

ROW_ID_x  SUBJECT_ID  HADM_ID            ADMITTIME            DISCHTIME DEATHTIME ADMISSION_TYPE         ADMISSION_LOCATION         DISCHARGE_LOCATION INSURANCE LANGUAGE      RELIGION MARITAL_STATUS ETHNICITY            EDREGTIME            EDOUTTIME                                          DIAGNOSIS  HOSPITAL_EXPIRE_FLAG  HAS_CHARTEVENTS_DATA  ROW_ID_y event_type  SEQ_NUM ICD9_CODE   ROW_ID               SHORT_TITLE                                         LONG_TITLE SEQ_NUM_str                         CLINICAL_RECORD                                      RECORD_VECTOR
0             21          22   165315  2196-04-09 12:26:00  2196-04-10 15:54:00       NaN      EMERGENCY       EMERGENCY ROOM ADMIT  DISC-TRAN CANCER/CHLDRN H   Private      NaN  UNOBTAINABLE        MARRIED     WHITE  2196-04-09 10:06:00  2196-04-09 13:24:00                            BENZODIAZEPINE OVERDOSE                     0                     1    9894.0  treatment      1.0      9604   3607.0  Insert endotracheal tube                     Insertion of endotracheal tube         1.0  treatment Insert endotracheal tube 1.0  [-0.8279435, 0.17752808, 0.8706687, 0.08610045...
1             21          22   165315  2196-04-09 12:26:00  2196-04-10 15:54:00       NaN      EMERGENCY       EMERGENCY ROOM ADMIT  DISC-TRAN CANCER/CHLDRN H   Private      NaN  UNOBTAINABLE        MARRIED     WHITE  2196-04-09 10:06:00  2196-04-09 13:24:00                            BENZODIAZEPINE OVERDOSE                     0                     1    9895.0  treatment      2.0      9671   3678.0  Cont inv mec ven <96 hrs  Continuous invasive mechanical ventilation for...         2.0  treatment Cont inv mec ven <96 hrs 2.0  [-0.7221991, 0.54226786, 0.8381936, -0.5248425...
2             21          22   165315  2196-04-09 12:26:00  2196-04-10 15:54:00       NaN      EMERGENCY       EMERGENCY ROOM ADMIT  DISC-TRAN CANCER/CHLDRN H   Private      NaN  UNOBTAINABLE        MARRIED     WHITE  2196-04-09 10:06:00  2196-04-09 13:24:00                            BENZODIAZEPINE OVERDOSE                     0                     1    9896.0  treatment      3.0      9634   3632.0  Gastric tube irrigat NEC            Other irrigation of (naso-)gastric tube         3.0  treatment Gastric tube irrigat NEC 3.0  [-0.36163458, 0.5326249, 0.35267824, -0.482253...
3             22          23   152223  2153-09-03 07:15:00  2153-09-08 19:10:00       NaN       ELECTIVE  PHYS REFERRAL/NORMAL DELI           HOME HEALTH CARE  Medicare      NaN      CATHOLIC        MARRIED     WHITE                  NaN                  NaN  CORONARY ARTERY DISEASE\CORONARY ARTERY BYPASS...                     0                     1  105252.0  treatment      1.0      3613    663.0  Aortocor bypas-3 cor art  (Aorto)coronary bypass of three coronary arteries         1.0  treatment Aortocor bypas-3 cor art 1.0  [-0.21862356, 0.36710623, -0.43862176, -0.2388...
4             22          23   152223  2153-09-03 07:15:00  2153-09-08 19:10:00       NaN       ELECTIVE  PHYS REFERRAL/NORMAL DELI           HOME HEALTH CARE  Medicare      NaN      CATHOLIC        MARRIED     WHITE                  NaN                  NaN  CORONARY ARTERY DISEASE\CORONARY ARTERY BYPASS...                     0                     1  105253.0  treatment      2.0      3615    665.0  1 int mam-cor art bypass     Single internal mammary-coronary artery bypass         2.0  treatment 1 int mam-cor art bypass 2.0  [0.18019682, 0.60992765, -1.0317271, -0.872899...
...          ...         ...      ...                  ...                  ...       ...            ...                        ...                        ...       ...      ...           ...            ...       ...                  ...                  ...                                                ...                   ...                   ...       ...        ...      ...       ...      ...                       ...                                                ...         ...                   

In [187]:
df.to_csv (r'/Users/liloli/Desktop/6895Project/Carepre/admission_treatment.csv', index = False, header=True)

In [50]:
df.shape

(905626, 29)

For data-processing convenience, convert the dataframe df to a list of dictionarys. Each dictionary in the list represents 1 row, where column name is the key.

In [51]:
l=list()
for row in df.iterrows(): 
    d = row[1].to_dict()
    l.append(d)

In [52]:
len(l)

905626

In [53]:
l[:5]

[{'ROW_ID_x': 21,
  'SUBJECT_ID': 22,
  'HADM_ID': 165315,
  'ADMITTIME': '2196-04-09 12:26:00',
  'DISCHTIME': '2196-04-10 15:54:00',
  'DEATHTIME': nan,
  'ADMISSION_TYPE': 'EMERGENCY',
  'ADMISSION_LOCATION': 'EMERGENCY ROOM ADMIT',
  'DISCHARGE_LOCATION': 'DISC-TRAN CANCER/CHLDRN H',
  'INSURANCE': 'Private',
  'LANGUAGE': nan,
  'RELIGION': 'UNOBTAINABLE',
  'MARITAL_STATUS': 'MARRIED',
  'ETHNICITY': 'WHITE',
  'EDREGTIME': '2196-04-09 10:06:00',
  'EDOUTTIME': '2196-04-09 13:24:00',
  'DIAGNOSIS': 'BENZODIAZEPINE OVERDOSE',
  'HOSPITAL_EXPIRE_FLAG': 0,
  'HAS_CHARTEVENTS_DATA': 1,
  'ROW_ID_y': 9894.0,
  'event_type': 'treatment',
  'SEQ_NUM': 1.0,
  'ICD9_CODE': 9604.0,
  'ROW_ID': 3607.0,
  'SHORT_TITLE': 'Insert endotracheal tube',
  'LONG_TITLE': 'Insertion of endotracheal tube',
  'SEQ_NUM_str': '1.0',
  'CLINICAL_RECORD': 'treatment Insert endotracheal tube 1.0',
  'RECORD_VECTOR': array([-0.8279435 ,  0.17752808,  0.8706687 ,  0.08610045, -0.02621948,
          1.5901188 

In [70]:
import numpy

Since each unique hospital visit for a patient is assigned a unique HADM_ID, the ADMISSIONS table can be considered as a definition table for HADM_ID. It is possible for this table to have duplicate SUBJECT_ID, indicating that a single patient had multiple admissions to the hospital. However, for each hospital visit, there could be multiple treatments. Thus, there are multiple rows for one HADM_ID. We want to map each hospital visit (each unique HADM_ID) to a list of records, which can be seen as a event-matrix.

use numpy.mean to center the matrix.

In [71]:
event_seq = dict()
for row in l:
    hadm_id = row['HADM_ID']
    if hadm_id not in event_seq.keys():
        event_seq[hadm_id] = list()
        event_seq[hadm_id].append(numpy.mean(row['RECORD_VECTOR']))
    else:
        event_seq[hadm_id].append(numpy.mean(row['RECORD_VECTOR']))

For example: when HADM_ID=100001, there are seven treatments.

In [72]:
event_seq[100001]

[0.0060770703,
 0.14726053,
 0.022220626,
 0.018554159,
 0.026176529,
 0.06426576,
 0.060381535,
 0.08051553,
 0.029220538,
 0.04516913,
 0.0961397,
 0.10563836,
 0.042630706,
 0.061899316,
 0.0075552543,
 0.017032066,
 0.016641878]

In [73]:
event_seq[100003]

[0.118318684,
 0.08475308,
 0.10119746,
 0.14092293,
 0.036108088,
 -0.054012507,
 0.05954881,
 0.006155901,
 0.03863813,
 0.03944859,
 -0.0046044015,
 0.063674904,
 0.04165874]

In [74]:
len(event_seq)

58976

In [75]:
len(df)

905626

There 660185 treatment/diagnose records in the dataframe and 58976 unique hospital visit.

Further clean df:

In [76]:
df_event_matrix = df.drop_duplicates(['HADM_ID'])[['SUBJECT_ID','HADM_ID','event_type','ADMISSION_LOCATION','SEQ_NUM','ICD9_CODE','ADMISSION_TYPE','ADMITTIME','DEATHTIME','DIAGNOSIS','DISCHARGE_LOCATION','DISCHTIME','EDOUTTIME','EDREGTIME','ETHNICITY', 'HAS_CHARTEVENTS_DATA','HOSPITAL_EXPIRE_FLAG', 'INSURANCE','LANGUAGE','LONG_TITLE','MARITAL_STATUS']]
df_event_matrix

SUBJECT_ID  HADM_ID event_type         ADMISSION_LOCATION  SEQ_NUM ICD9_CODE ADMISSION_TYPE            ADMITTIME            DEATHTIME                                          DIAGNOSIS         DISCHARGE_LOCATION            DISCHTIME            EDOUTTIME            EDREGTIME ETHNICITY  HAS_CHARTEVENTS_DATA  HOSPITAL_EXPIRE_FLAG INSURANCE LANGUAGE                                         LONG_TITLE MARITAL_STATUS
0               22   165315  treatment       EMERGENCY ROOM ADMIT      1.0      9604      EMERGENCY  2196-04-09 12:26:00                  NaN                            BENZODIAZEPINE OVERDOSE  DISC-TRAN CANCER/CHLDRN H  2196-04-10 15:54:00  2196-04-09 13:24:00  2196-04-09 10:06:00     WHITE                     1                     0   Private      NaN                     Insertion of endotracheal tube        MARRIED
3               23   152223  treatment  PHYS REFERRAL/NORMAL DELI      1.0      3613       ELECTIVE  2153-09-03 07:15:00                  NaN  CORONARY ARTERY DISEASE\CORONARY ARTERY BYPASS...           HOME HEALTH CARE  2153-09-08 19:10:00                  NaN                  NaN     WHITE                     1                     0  Medicare      NaN  (Aorto)coronary bypass of three coronary arteries        MARRIED
10              23   124321  treatment  TRANSFER FROM HOSP/EXTRAM      1.0       151      EMERGENCY  2157-10-18 19:34:00                  NaN                                         BRAIN MASS           HOME HEALTH CARE  2157-10-25 14:00:00                  NaN                  NaN     WHITE                     1                     0  Medicare     ENGL  Excision of lesion or tissue of cerebral meninges        MARRIED
11              24   161859  treatment  TRANSFER FROM HOSP/EXTRAM      1.0      3601      EMERGENCY  2139-06-06 16:14:00                  NaN                     INTERIOR MYOCARDIAL INFARCTION                       HOME  2139-06-09 12:48:00                  NaN                  NaN     WHITE                     1                     0   Private      NaN                                                NaN         SINGLE
17              25   129635  treatment       EMERGENCY ROOM ADMIT      1.0      3601      EMERGENCY  2160-11-02 02:06:00                  NaN                            ACUTE CORONARY SYNDROME                       HOME  2160-11-05 14:55:00  2160-11-02 04:27:00  2160-11-02 01:01:00     WHITE                     1                     0   Private      NaN                                                NaN        MARRIED
...            ...      ...        ...                        ...      ...       ...            ...                  ...                  ...                                                ...                        ...                  ...                  ...                  ...       ...                   ...                   ...       ...      ...                                                ...            ...
254564       98800   191113  treatment  CLINIC REFERRAL/PREMATURE      1.0       881      EMERGENCY  2131-03-30 21:13:00                  NaN                                             TRAUMA                       HOME  2131-04-02 15:02:00  2131-03-30 22:41:00  2131-03-30 19:44:00     WHITE                     1                     0   Private     ENGL   Linear repair of laceration of eyelid or eyebrow         SINGLE
254568       98802   101071  treatment  CLINIC REFERRAL/PREMATURE      1.0      9671      EMERGENCY  2151-03-05 20:00:00  2151-03-06 09:10:00                                                SAH               DEAD/EXPIRED  2151-03-06 09:10:00  2151-03-05 21:06:00  2151-03-05 17:23:00     WHITE                     1                     1  Medicare     ENGL  Continuous invasive mechanical ventilation for...        WIDOWED
254569       98805   122631  treatment  PHYS REFERRAL/NORMAL DELI      1.0      4576       ELECTIVE  2200-09-12 07:15:00                  NaN                                   RENAL CANCER/SDA           HOME HE

In [77]:
event_df = pd.Series(event_seq, name='EVENT_MATRIX')
event_df.head()

165315    [0.30737028, 0.14932567, 0.031850547, 0.029994...
152223    [0.022115398, 0.041287336, 0.061276045, 0.1448...
124321    [-0.014348476, 0.009678095, 0.011745515, 0.027...
161859    [0.0060770703, 0.076156534, 0.065352686, -0.00...
129635    [0.0060770703, 0.076156534, 0.049859148, 0.058...
Name: EVENT_MATRIX, dtype: object

In [78]:
event_seq

{165315: [0.30737028,
  0.14932567,
  0.031850547,
  0.029994622,
  0.020776283,
  0.014137905,
  0.054572076,
  0.0060770703,
  0.04954997,
  -0.0046044015],
 152223: [0.022115398,
  0.041287336,
  0.061276045,
  0.1448137,
  0.08690523,
  -0.0021725309,
  0.03159981,
  -0.054061614,
  -0.013469665,
  0.07458045,
  -0.005825422,
  -0.043863665,
  -0.023253866,
  0.009030235,
  0.019089794],
 124321: [-0.014348476,
  0.009678095,
  0.011745515,
  0.027689518,
  0.07445771,
  -0.033028692,
  -0.10272802,
  -0.010570187,
  0.045938283,
  0.016927863,
  -0.06313104],
 161859: [0.0060770703,
  0.076156534,
  0.065352686,
  -0.005870265,
  0.05398252,
  0.04279449,
  0.020696063,
  -0.063599885,
  -0.012567502,
  -0.008556554],
 129635: [0.0060770703,
  0.076156534,
  0.049859148,
  0.05874746,
  0.05289717,
  0.014983666,
  0.04292972,
  -0.018547725,
  0.016475413,
  -0.07078017,
  0.037480526,
  -0.039721996,
  -0.043492425],
 197661: [-0.0364666,
  0.011263881,
  0.030014815,
  -0.05653

把同一个patient的不同head_id放在一起形成一个新的matrix。使用dtw方法。

In [79]:
event_df.index.name = 'HADM_ID'
event_df = event_df.reset_index()
event_df.head()

HADM_ID                                       EVENT_MATRIX
0   165315  [0.30737028, 0.14932567, 0.031850547, 0.029994...
1   152223  [0.022115398, 0.041287336, 0.061276045, 0.1448...
2   124321  [-0.014348476, 0.009678095, 0.011745515, 0.027...
3   161859  [0.0060770703, 0.076156534, 0.065352686, -0.00...
4   129635  [0.0060770703, 0.076156534, 0.049859148, 0.058...

Now, we have the EVENT_MATRIX for each HADM_ID(one unique visit to hospital). 
Merge infomation table `df_event_matrix` of one visit(unique HADM_ID) and __embedding event matrix__ `event_df`

In [80]:
data = pd.merge(df_event_matrix, event_df, on=["HADM_ID"], how='left')
data

SUBJECT_ID  HADM_ID event_type         ADMISSION_LOCATION  SEQ_NUM ICD9_CODE ADMISSION_TYPE            ADMITTIME            DEATHTIME                                          DIAGNOSIS         DISCHARGE_LOCATION            DISCHTIME            EDOUTTIME            EDREGTIME ETHNICITY  HAS_CHARTEVENTS_DATA  HOSPITAL_EXPIRE_FLAG INSURANCE LANGUAGE                                         LONG_TITLE MARITAL_STATUS                                       EVENT_MATRIX
0              22   165315  treatment       EMERGENCY ROOM ADMIT      1.0      9604      EMERGENCY  2196-04-09 12:26:00                  NaN                            BENZODIAZEPINE OVERDOSE  DISC-TRAN CANCER/CHLDRN H  2196-04-10 15:54:00  2196-04-09 13:24:00  2196-04-09 10:06:00     WHITE                     1                     0   Private      NaN                     Insertion of endotracheal tube        MARRIED  [0.30737028, 0.14932567, 0.031850547, 0.029994...
1              23   152223  treatment  PHYS REFERRAL/NORMAL DELI      1.0      3613       ELECTIVE  2153-09-03 07:15:00                  NaN  CORONARY ARTERY DISEASE\CORONARY ARTERY BYPASS...           HOME HEALTH CARE  2153-09-08 19:10:00                  NaN                  NaN     WHITE                     1                     0  Medicare      NaN  (Aorto)coronary bypass of three coronary arteries        MARRIED  [0.022115398, 0.041287336, 0.061276045, 0.1448...
2              23   124321  treatment  TRANSFER FROM HOSP/EXTRAM      1.0       151      EMERGENCY  2157-10-18 19:34:00                  NaN                                         BRAIN MASS           HOME HEALTH CARE  2157-10-25 14:00:00                  NaN                  NaN     WHITE                     1                     0  Medicare     ENGL  Excision of lesion or tissue of cerebral meninges        MARRIED  [-0.014348476, 0.009678095, 0.011745515, 0.027...
3              24   161859  treatment  TRANSFER FROM HOSP/EXTRAM      1.0      3601      EMERGENCY  2139-06-06 16:14:00                  NaN                     INTERIOR MYOCARDIAL INFARCTION                       HOME  2139-06-09 12:48:00                  NaN                  NaN     WHITE                     1                     0   Private      NaN                                                NaN         SINGLE  [0.0060770703, 0.076156534, 0.065352686, -0.00...
4              25   129635  treatment       EMERGENCY ROOM ADMIT      1.0      3601      EMERGENCY  2160-11-02 02:06:00                  NaN                            ACUTE CORONARY SYNDROME                       HOME  2160-11-05 14:55:00  2160-11-02 04:27:00  2160-11-02 01:01:00     WHITE                     1                     0   Private      NaN                                                NaN        MARRIED  [0.0060770703, 0.076156534, 0.049859148, 0.058...
...           ...      ...        ...                        ...      ...       ...            ...                  ...                  ...                                                ...                        ...                  ...                  ...                  ...       ...                   ...                   ...       ...      ...                                                ...            ...                                                ...
58971       98800   191113  treatment  CLINIC REFERRAL/PREMATURE      1.0       881      EMERGENCY  2131-03-30 21:13:00                  NaN                                             TRAUMA                       HOME  2131-04-02 15:02:00  2131-03-30 22:41:00  2131-03-30 19:44:00     WHITE                     1                     0   Private     ENGL   Linear repair of laceration of eyelid or eyebrow         SINGLE  [0.020543395, 0.003161966, 0.051592965, 0.0497...
58972       98802   101071  treatment  CLINIC REFERRAL/PREMATURE      1.0      9671      EMERGENCY  2151-03-05 20:00:00  2151-03-06 09:10:00                                                SAH               DEAD/EXPIRED  2151-03-06 

In [81]:
data.shape

(58976, 22)

In [149]:
patient_id = data.drop_duplicates(['SUBJECT_ID'])[['SUBJECT_ID']]

SUBJECT_ID
0              22
1              23
3              24
4              25
5              26
...           ...
58969       98794
58970       98797
58971       98800
58972       98802
58973       98805

[46520 rows x 1 columns]

According to the calculus, there are 46520 unique patients with 58976 hospital visit records, which means one patient can visit the hospital for multiple times.

Data-frame to list

In [90]:
l2=list()
for row in data.iterrows(): 
    d = row[1].to_dict()
    l2.append(d)

In [91]:
l2[:5]

[{'SUBJECT_ID': 22,
  'HADM_ID': 165315,
  'event_type': 'treatment',
  'ADMISSION_LOCATION': 'EMERGENCY ROOM ADMIT',
  'SEQ_NUM': 1.0,
  'ICD9_CODE': 9604.0,
  'ADMISSION_TYPE': 'EMERGENCY',
  'ADMITTIME': '2196-04-09 12:26:00',
  'DEATHTIME': nan,
  'DIAGNOSIS': 'BENZODIAZEPINE OVERDOSE',
  'DISCHARGE_LOCATION': 'DISC-TRAN CANCER/CHLDRN H',
  'DISCHTIME': '2196-04-10 15:54:00',
  'EDOUTTIME': '2196-04-09 13:24:00',
  'EDREGTIME': '2196-04-09 10:06:00',
  'ETHNICITY': 'WHITE',
  'HAS_CHARTEVENTS_DATA': 1,
  'HOSPITAL_EXPIRE_FLAG': 0,
  'INSURANCE': 'Private',
  'LANGUAGE': nan,
  'LONG_TITLE': 'Insertion of endotracheal tube',
  'MARITAL_STATUS': 'MARRIED',
  'EVENT_MATRIX': [0.30737028,
   0.14932567,
   0.031850547,
   0.029994622,
   0.020776283,
   0.014137905,
   0.054572076,
   0.0060770703,
   0.04954997,
   -0.0046044015]},
 {'SUBJECT_ID': 23,
  'HADM_ID': 152223,
  'event_type': 'treatment',
  'ADMISSION_LOCATION': 'PHYS REFERRAL/NORMAL DELI',
  'SEQ_NUM': 1.0,
  'ICD9_CODE':

In [280]:
patient_seq = dict()
for row in l2:
    patient_id = row['SUBJECT_ID']
    if patient_id not in patient_seq.keys():
        patient_seq[patient_id] = list()
        patient_seq[patient_id].append(np.array(row['EVENT_MATRIX']))
    else:
        patient_seq[patient_id].append(np.array(row['EVENT_MATRIX']))

In [281]:
patient_seq

{22: [array([ 0.30737028,  0.14932567,  0.03185055,  0.02999462,  0.02077628,
          0.01413791,  0.05457208,  0.00607707,  0.04954997, -0.0046044 ],
        dtype=float32)],
 23: [array([ 0.0221154 ,  0.04128734,  0.06127604,  0.1448137 ,  0.08690523,
         -0.00217253,  0.03159981, -0.05406161, -0.01346967,  0.07458045,
         -0.00582542, -0.04386367, -0.02325387,  0.00903023,  0.01908979],
        dtype=float32),
  array([-0.01434848,  0.00967809,  0.01174552,  0.02768952,  0.07445771,
         -0.03302869, -0.10272802, -0.01057019,  0.04593828,  0.01692786,
         -0.06313104], dtype=float32)],
 24: [array([ 0.00607707,  0.07615653,  0.06535269, -0.00587026,  0.05398252,
          0.04279449,  0.02069606, -0.06359988, -0.0125675 , -0.00855655],
        dtype=float32)],
 25: [array([ 0.00607707,  0.07615653,  0.04985915,  0.05874746,  0.05289717,
          0.01498367,  0.04292972, -0.01854772,  0.01647541, -0.07078017,
          0.03748053, -0.039722  , -0.04349243], dtyp

In [282]:
patient_df = pd.Series(patient_seq, name='patient_vector')
patient_df.head()

22    [[0.30737028, 0.14932567, 0.031850547, 0.02999...
23    [[0.022115398, 0.041287336, 0.061276045, 0.144...
24    [[0.0060770703, 0.076156534, 0.065352686, -0.0...
25    [[0.0060770703, 0.076156534, 0.049859148, 0.05...
26    [[-0.0364666, 0.011263881, 0.030014815, -0.056...
Name: patient_vector, dtype: object

现在每个patient的record都被数字化了。为了能对每个patient进行比较，使用dynamic time wrapping。

In [142]:
patient_matrix = pd.Series(patient_seq, name='patient_array')
patient_matrix.index.name = 'SUBJECT_ID'
patient_matrix = patient_matrix.reset_index()
patient_matrix.head()

SUBJECT_ID                                      patient_array
0          22  [[0.30737028, 0.14932567, 0.031850547, 0.02999...
1          23  [[0.022115398, 0.041287336, 0.061276045, 0.144...
2          24  [[0.0060770703, 0.076156534, 0.065352686, -0.0...
3          25  [[0.0060770703, 0.076156534, 0.049859148, 0.05...
4          26  [[-0.0364666, 0.011263881, 0.030014815, -0.056...

In [137]:
patient_info = df_event_matrix.drop_duplicates(['SUBJECT_ID'])[['SUBJECT_ID','ETHNICITY', 'HAS_CHARTEVENTS_DATA','HOSPITAL_EXPIRE_FLAG', 'INSURANCE','LANGUAGE','MARITAL_STATUS']]

In [150]:
patient_info

SUBJECT_ID              ETHNICITY  HAS_CHARTEVENTS_DATA  HOSPITAL_EXPIRE_FLAG INSURANCE LANGUAGE MARITAL_STATUS
0               22                  WHITE                     1                     0   Private      NaN        MARRIED
3               23                  WHITE                     1                     0  Medicare      NaN        MARRIED
11              24                  WHITE                     1                     0   Private      NaN         SINGLE
17              25                  WHITE                     1                     0   Private      NaN        MARRIED
26              26  UNKNOWN/NOT SPECIFIED                     1                     0  Medicare      NaN         SINGLE
...            ...                    ...                   ...                   ...       ...      ...            ...
254560       98794                  WHITE                     1                     0  Medicare     ENGL        MARRIED
254562       98797                  WHITE                     1                     1  Medicare     ENGL        WIDOWED
254564       98800                  WHITE                     1                     0   Private     ENGL         SINGLE
254568       98802                  WHITE                     1                     1  Medicare     ENGL        WIDOWED
254569       98805                  WHITE                     1                     0   Private     ENGL        MARRIED

[46520 rows x 7 columns]

In [234]:
patient_seq

{22: [array([ 0.30737028,  0.14932567,  0.03185055,  0.02999462,  0.02077628,
          0.01413791,  0.05457208,  0.00607707,  0.04954997, -0.0046044 ],
        dtype=float32)],
 23: [array([ 0.0221154 ,  0.04128734,  0.06127604,  0.1448137 ,  0.08690523,
         -0.00217253,  0.03159981, -0.05406161, -0.01346967,  0.07458045,
         -0.00582542, -0.04386367, -0.02325387,  0.00903023,  0.01908979],
        dtype=float32),
  array([-0.01434848,  0.00967809,  0.01174552,  0.02768952,  0.07445771,
         -0.03302869, -0.10272802, -0.01057019,  0.04593828,  0.01692786,
         -0.06313104], dtype=float32)],
 24: [array([ 0.00607707,  0.07615653,  0.06535269, -0.00587026,  0.05398252,
          0.04279449,  0.02069606, -0.06359988, -0.0125675 , -0.00855655],
        dtype=float32)],
 25: [array([ 0.00607707,  0.07615653,  0.04985915,  0.05874746,  0.05289717,
          0.01498367,  0.04292972, -0.01854772,  0.01647541, -0.07078017,
          0.03748053, -0.039722  , -0.04349243], dtyp

In [287]:
def max_length(x):
    max_length = 0
    for i in range(0,len(x)):
        if(len(x[i])>max_length):
            max_length = len(x[i])
    return(max_length)
            

total_max = 0
for key in patient_seq:
    item = patient_seq[key]
    length = len(patient_seq[key])
    if(length>1):
        ml = max_length(item)
        if(ml>total_max):
            total_max = ml
            
#Prepare for dynamic time wrapping
for key in patient_seq:
    item = patient_seq[key]
    length = len(patient_seq[key])
    if(length>-1):
        for i in range(0, len(item)):
            gap = 100  - len(item[i])
            if(gap>0):
                arr_append = np.repeat(0,gap)
                item[i] = np.append(item[i] , arr_append)

In [288]:
#check whether every patient's matrix is valid
for key in patient_seq:
    item = patient_seq[key]
    length = len(patient_seq[key])
    if(length>1):
        for i in range(0, len(item)):
            gap = 100  - len(item[i])
            if(gap>0):
                print("Something goes wrong")
print("correct")

correct


In [289]:
import numpy as np
from scipy.spatial.distance import euclidean
from fastdtw import fastdtw

In [290]:
x = patient_seq[22]
y = patient_seq[23]
distance, path = fastdtw(x, y, dist=euclidean)
distance

0.7567475169441176

In [292]:
for key in patient_seq:
    item = patient_seq[key]
    n = len(item)
    if(n==1):
        if(len(item[0])!=100):
            print("wrong")
            print(key) 
    else:
        for i in range(0, n):
            if(len(item[i])!=100):
                print("wrong")
                print(key)


In [328]:
len(patient_seq)

46520

In [295]:
#calculate distance from a given patient to others:
def calculate_dist(patient_id):
    patient_dist = dict()
    x = patient_seq[patient_id]
    for key in patient_seq:
        y= patient_seq[key]
        distance, path = fastdtw(x, y, dist=euclidean)
        patient_dist[key]=distance
    return(patient_dist)
   
            
calculate_dist(22)

{22: 0.0,
 23: 0.7567475169441176,
 24: 0.3321754313457694,
 25: 0.33076293862107253,
 26: 0.4247068062762086,
 27: 0.2067174093047957,
 28: 0.45485196379645526,
 30: 0.3548935623376853,
 31: 0.42181399065482955,
 32: 0.38001799613274795,
 33: 0.32028270542517395,
 34: 0.6306943573776755,
 35: 0.3548962865528676,
 36: 1.1003249671502633,
 37: 0.381238740671879,
 38: 0.42350217389295725,
 39: 0.3779298569191118,
 41: 0.37508634595280205,
 357: 2.010479183798969,
 358: 0.17200235162357863,
 359: 0.341191127451951,
 360: 0.27839983111822847,
 361: 0.523358113559478,
 362: 0.5820258342095936,
 363: 0.1538637590372055,
 364: 0.4385547762578525,
 365: 0.3062695733784216,
 366: 0.35348052710165073,
 367: 0.2900591575993211,
 368: 0.7590989081070572,
 369: 0.319792667503448,
 370: 0.36436430186636115,
 371: 0.33373470561416246,
 372: 0.25549235000178894,
 42: 0.3978888740616372,
 43: 0.4614921954808643,
 44: 0.34138011412454894,
 45: 0.36366072539351213,
 46: 0.34238934085903083,
 49: 0.354430

In [ ]:
import operator
x = calculate_dist(22)
sorted_x = sorted(x.items(), key=operator.itemgetter(1))
sorted_x

In [332]:
similar_dict = dict()
i=0
for key in patient_seq:
    x = calculate_dist(key)
    sorted_x = sorted(x.items(), key=operator.itemgetter(1))
    record = list()
    record.append(sorted_x[1][0])
    record.append(sorted_x[1][1])
    record.append(sorted_x[2][0])
    record.append(sorted_x[2][1])
    record.append(sorted_x[3][0])
    record.append(sorted_x[3][1]) 
    similar_dict[key] = record  
    i=i+1
    if(i>1000):
        break

In [336]:
similar_dict

{22: [6391,
  0.042710509140304886,
  4450,
  0.050738384083492134,
  12992,
  0.05208191909157845],
 23: [58263,
  0.28318681828326864,
  9643,
  0.2884666474278593,
  9786,
  0.2912980771514181],
 24: [12168,
  0.012070365737787027,
  13681,
  0.021603933917212048,
  2536,
  0.02775853974498135],
 25: [12448,
  0.09009700251928231,
  456,
  0.0903843692116209,
  21533,
  0.09439017978988763],
 26: [64759,
  0.1001838828776069,
  22184,
  0.10068855379088665,
  29724,
  0.10210571075236975],
 27: [3291, 0.0, 11103, 0.0, 31997, 0.0],
 28: [27438,
  0.16465306565816304,
  27244,
  0.17963127371198734,
  69673,
  0.18208866751442881],
 30: [16858,
  0.079456460499609,
  19261,
  0.08149787192340426,
  6338,
  0.08346589168042887],
 31: [17377,
  0.12175980206930187,
  31250,
  0.1445992672284439,
  9269,
  0.1454766350743348],
 32: [11272,
  0.08436851920458846,
  26346,
  0.08854110316796557,
  1956,
  0.09114868208496546],
 33: [11174,
  0.03310361514437943,
  2992,
  0.037062731911117

In [347]:
similarity_df = pd.DataFrame.from_dict(similar_dict, orient='index', columns=['top1_id', 'top1_dist', 'top2_id', 'top2_dist','top3_id','top3_dist'])

In [352]:
i=0
new_col = list()
for key in patient_seq:
    new_col.append(key)
    i=i+1
    if(i>1000):
        break
similarity_df.insert(0, "SUBJECT_ID",new_col , True)
similarity_df

SUBJECT_ID  top1_id  top1_dist  top2_id  top2_dist  top3_id  top3_dist
22            22     6391   0.042711     4450   0.050738    12992   0.052082
23            23    58263   0.283187     9643   0.288467     9786   0.291298
24            24    12168   0.012070    13681   0.021604     2536   0.027759
25            25    12448   0.090097      456   0.090384    21533   0.094390
26            26    64759   0.100184    22184   0.100689    29724   0.102106
...          ...      ...        ...      ...        ...      ...        ...
1466        1466    24319   0.187194    22734   0.187374     1765   0.204710
1467        1467     3394   0.111639    28937   0.113635    25740   0.115855
1469        1469    30103   0.113447    23160   0.117417    25630   0.128157
1470        1470    46946   0.212887    31144   0.215080    15090   0.216135
1471        1471    26079   0.118054    87907   0.120582     7333   0.135723

[1001 rows x 7 columns]

Download the dataframe and store it into a csv.

In [353]:
similarity_df.to_csv (r'/Users/liloli/Desktop/6895Project/Carepre/similarity.csv', index = False, header=True)

In [363]:
recommendation = similarity_df.merge(data , left_on='top1_id', right_on='SUBJECT_ID')

In [364]:
recommendation

SUBJECT_ID_x  top1_id  top1_dist  top2_id  top2_dist  top3_id  top3_dist  SUBJECT_ID_y  HADM_ID event_type         ADMISSION_LOCATION  SEQ_NUM ICD9_CODE ADMISSION_TYPE            ADMITTIME DEATHTIME                             DIAGNOSIS         DISCHARGE_LOCATION            DISCHTIME            EDOUTTIME            EDREGTIME              ETHNICITY  HAS_CHARTEVENTS_DATA  HOSPITAL_EXPIRE_FLAG   INSURANCE LANGUAGE                                         LONG_TITLE MARITAL_STATUS                                       EVENT_MATRIX
0               22     6391   0.042711     4450   0.050738    12992   0.052082          6391   119373  treatment       EMERGENCY ROOM ADMIT      1.0      9604      EMERGENCY  2170-05-07 21:30:00       NaN                              OVERDOSE  DISC-TRAN CANCER/CHLDRN H  2170-05-10 11:25:00  2170-05-07 23:07:00  2170-05-07 20:25:00                  WHITE                     1                     0  Government      NaN                     Insertion of endotracheal tube         SINGLE  [0.30737028, 0.14932567, 0.010712185, 0.032199...
1               23    58263   0.283187     9643   0.288467     9786   0.291298         58263   190790  treatment  TRANSFER FROM HOSP/EXTRAM      1.0      3782      EMERGENCY  2153-06-20 22:15:00       NaN  ACUTE MYOCARDIAL INFARCTION; AT. FIB                        SNF  2153-06-28 16:30:00                  NaN                  NaN                  WHITE                     1                     0    Medicare     ENGL  Initial insertion of single-chamber device, ra...        WIDOWED  [0.015437689, 0.013389931, 0.01377173, 0.05183...
2               24    12168   0.012070    13681   0.021604     2536   0.027759         12168   175288  treatment       EMERGENCY ROOM ADMIT      1.0      3601      EMERGENCY  2155-07-20 10:25:00       NaN                       CHEST PAIN/CATH                       HOME  2155-07-22 15:04:00                  NaN                  NaN                  WHITE                     1                     0    Medicare      NaN                                                NaN        MARRIED  [0.0060770703, 0.076156534, 0.065352686, -0.00...
3               25    12448   0.090097      456   0.090384    21533   0.094390         12448   164086  treatment  CLINIC REFERRAL/PREMATURE      1.0      3612      EMERGENCY  2109-12-16 11:46:00       NaN          CORONARY ARTERY DISEASE\CATH                       HOME  2109-12-23 13:45:00                  NaN                  NaN                  WHITE                     1                     0     Private      NaN    (Aorto)coronary bypass of two coronary arteries        MARRIED  [-0.0069909454, 0.07259793, 0.079707764, 0.033...
4               26    64759   0.100184    22184   0.100689    29724   0.102106         64759   177867  treatment  PHYS REFERRAL/NORMAL DELI      1.0      3259       ELECTIVE  2121-03-18 08:00:00       NaN                           LUNG CA/SDA                       HOME  2121-03-24 16:45:00                  NaN                  NaN                  WHITE                     1                     0     Private     ENGL                Other and unspecified pneumonectomy       DIVORCED  [-0.038951922, -0.012032778, -0.021025207, -0....
...            ...      ...        ...      ...        ...      ...        ...           ...      ...        ...                        ...      ...       ...            ...                  ...       ...                                   ...                        ...                  ...                  ...                  ...                    ...                   ...                   ...         ...      ...                                                ...            ...                                                ...
1124          1466    24319   0.187194    22734   0.187374     1765   0.204710         24319   159994  treatment  CLINIC REFERRAL/PREMATURE      1.0      9671        NEWBORN  2173-05-27 08:43:00       NaN                               NEWBO